### ISSUE TITLE
ensemble

### ISSUE THREAD

**Issue by flamby** [imoscovitz/wittgenstein#5]
Hi @imoscovitz ,

Suppose you trained several classifiers with different algorithms (IREP and RIPPER) with different feature sets and try to guess the best combination by stacking them with different weights.
Of course, one could simply retrain a classifier w/ all feature sets, but it can be computer intensive ; in fact, I tried that, since from an operational perspective, it's simpler to rely on one model than several, but the classifier w/ all features always performs worst than the ensemble combination I grid searched. Ensemble methods are quite often suggested to fight overfitting in the literature, and I plan to add recalibration to this ensemble experimentation.

Right now, iirc, since IREP or RIPPER use `dataframes` as input instead of `numpy` arrays, we can't rely on most ensemble meta-models which use the sklearn-way, i.e. numpy arrays.

So I came to do it manually like this:
```
probas = np.average(probas_list, axis=0, weights=weights)
```
and 
```
weights_list = itertools.product(ratios, repeat=len(probas_list))
```

This trick helped me gaining 2 points in precision, with only a limited ensemble grid search.

Also one could add to this {RIPPER,IREP} ensemble some `sklearn` algorithms like RandomForest, Bagging or DecisionTree, by computing their probabilities from the very same dataset.

Did you come up with an idea like this, perhaps more elegantly than mine?

**Comment by imoscovitz** [comment]
Hi @flamby,

Thanks for opening this issue -- these are interesting questions! Hadn't been too focused on ensembling these, so it's great to hear your feedback/thoughts.

> the classifier w/ all features always performs worst than the ensemble combination I grid searched

- That's very interesting. How are you selecting your feature subsets? Randomly, a la random forest? Systematically? Manually? 
- How many classifiers are you training for your bagging?
- With the first speed optimization committed, has bagging become more tractable?

> Of course, one could simply retrain a classifier w/ all feature sets, but it can be computer intensive

- What are you varying when you mention the possibility of retraining w/all feature sets?

> Right now, iirc, since IREP or RIPPER use dataframes as input instead of numpy arrays, we can't rely on most ensemble meta-models which use the sklearn-way, i.e. numpy arrays....Also one could add to this {RIPPER,IREP} ensemble some sklearn algorithms like RandomForest, Bagging or DecisionTree, by computing their probabilities from the very same dataset.

- Hadn't thought about this before, but with the new cn_optimize=True refactoring, it's become more practical to build in the option for passing in numpy training sets. I imagine that would be pretty good to include? Has not having that option felt limiting/ awkward?
- Would adding this be the ideal/most natural solution as a user? With allowing numpy training sets, would using sklearn ensemble techniques on top IREP/RIPPER become the same procedure as doing it with sklearn base classifiers?

Related:
- re grid-searching ```max_rules``` and ```max_total_conds```: I think trimming the ruleset after training should probably yield similar?/same? results as passing the parameters for fitting. (Not true of ```max_rule_conds``` though.) We could add a method for post-training that takes a dictionary of params and spits out different models, in order to allow for grid-searching those parameters without having to retrain. Might be called .trimmed or something. We'd have to decide, would be worth it, since the function is a little odd and might feel a little fiddly to use/document alongside searching other params, since it's not how grid-searching normally works?

Thanks!
Ilan

**Comment by flamby** [comment]
Hi @imoscovitz 

> > the classifier w/ all features always performs worst than the ensemble combination I grid searched
> 
> * That's very interesting. 
> How are you selecting your feature subsets? Randomly, a la random forest? Systematically? Manually?

I'm not selecting them beforehand. I let IREP doing it, since it turns out - for my use-case at least - to be a wonderful features selection method with highly correlated features, and better than tree-based algorithms at keeping a decent recall.

For instance, I generate dozens of feature sets with 1 or 2k+ features each, and train IREP classifiers on those sets. From the classifiers generated rules, I do a little domain-knowledge analysis of the features used in the rules to doublecheck that a feature is not used in a silly way ; if so, I exclude it from the next step :
Then I train another IREP classifier with the union of those selected features.
In the end, I usually have a very good classifier. Let's say that I end up with 10 classifiers trained w/ 10 different domains' features, doing an ensemble of them could make sense, since one classifier alone with their features' union does not outperform any classifiers alone (more on that below).

Regarding this homemade ensemble, I'll try to clarify with the below table. Having read again this original issue's post, I should have been more specific.

__Legend__ : 
* Name : classifier name, just to ease understanding
* Features set : description of the set of features used
* Features (before) : number of features in the initial dataframe
* Features (after) : the features used in the generated rules, after training
* Precision : the one returned by `predict`
* Best precision : the one return through `predict_proba` by adjusting decision threshold
  * again, this is specific to my use-case where precision is **the** metric and lower recall is just missed opportunities
* Recall : the recall associated to the best precision
* Comment

| Name | Features Set | Features (before) | Features (used in rules) | Precision | Best precision  | Recall | Comment |
| --- | --- | --- | --- | --- | --- | --- | --- | 
| IREP_1 | **A** | 1140 | 12 | 0.63 | 0.77 | 0.58 | - |
| IREP_2 | **B** | 971 | 9 | 0.63 | 0.80 | 0.33 | - |
| IREP_3 | **A+B** (after) | 21 | 18 | 0.65 | 0.76 | 0.4 | usage of features (_after_) of the first 2 classifiers |
| _homemade_ ensemble | **A+B** (after) | 21 | - | 0.66 | 0.82 | 0.32 | usage of features (_after_) of the first 2 classifiers & grid search of best weights for probabilities of classifiers IREP_1 and IREP_2

So as you see, the homemade ensemble, thanks to the weights which were offline gridsearched (so to speak) outperformed IREP_3, which makes sense in a way since it's a kind of selection bias. 

But what's more surprising is that _IREP 3_ does not do better than IREP_1 or IREP_2 alone ; whereas IREP_3 uses the union of features used in IREP 1 & 2 rules. If I changed the algorithm to RIPPER, with A+B features set as well, it is also not better.

> * How many classifiers are you training for your bagging?

I tried with up to 10 for now, with a very limited weights ratio combinations.
Since this ensemble ratio combination is offline (pure numpy), it's very quick to heuristically test for thousand of combinations (hence the selection bias ; but in my case I'll be able to mitigate that by validating against dozens of unseen datasets, averaging it so to speak). 
If you're familiar with [VotingClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.VotingClassifier.html) from sklearn, it's similar, but offline, as long as you stored the classifiers' probabilities and you used the very same dataset as input (even if they don't share the same features set ; the dataset must have the features of all of them).

> * With the first speed optimization committed, has bagging become more tractable?

With the huge IREP perf improvement, I'll be able to test more features set. So, in a sense, I'll be able to try this offline ensembling a lot more. But in my case, speed optimization is more useful for reducing drastically training time after having done a huge feature engineering (1k+ features).  

> > Of course, one could simply retrain a classifier w/ all feature sets, but it can be computer intensive
> 
> * What are you varying when you mention the possibility of retraining w/all feature sets?

I hope I clarified in my above table. English is not my mother tongue, so I have sometimes difficulties to express my thoughts ;-)

> > Right now, iirc, since IREP or RIPPER use dataframes as input instead of numpy arrays, we can't rely on most ensemble meta-models which use the sklearn-way, i.e. numpy arrays....Also one could add to this {RIPPER,IREP} ensemble some sklearn algorithms like RandomForest, Bagging or DecisionTree, by computing their probabilities from the very same dataset.
> 
> * Hadn't thought about this before, but with the new cn_optimize=True refactoring, it's become more practical to build in the option for passing in numpy training sets. I imagine that would be pretty good to include? Has not having that option felt limiting/ awkward?

Not a scikit-learn expert at all, just regular ML newbie, I see contradictory information on that sklearn prerequisite. See for instance [Building Scikit-Learn Pipelines With Pandas DataFrames
](https://ramhiser.com/post/2018-04-16-building-scikit-learn-pipeline-with-pandas-dataframe/).

> * Would adding this be the ideal/most natural solution as a user? With allowing numpy training sets, would using sklearn ensemble techniques on top IREP/RIPPER become the same procedure as doing it with sklearn base classifiers?

I think so. Sklearn is supposed to be loosely-coupled, and some articles I read on that let me think that it should not be a big deal then to use something like VotingClassifier to dynamically link IREP/RIPPER to other sklearn-compatible algorithms, to use it in a [pipeline](https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html) for instance.

> Related:
> 
> * re grid-searching `max_rules` and `max_total_conds`: I think trimming the ruleset after training should probably yield similar?/same? results as passing the parameters for fitting. (Not true of `max_rule_conds` though.) We could add a method for post-training that takes a dictionary of params and spits out different models, in order to allow for grid-searching those parameters without having to retrain. Might be called .trimmed or something. We'd have to decide, would be worth it, since the function is a little odd and might feel a little fiddly to use/document alongside searching other params, since it's not how grid-searching normally works?

In my case, since the generated rules are very simple (30 rules in the worst-case scenario), I don't think it will be useful. But I guess for NLP for instance it's completely different and it could be useful for sure.

Thanks

**Comment by flamby** [comment]
I'll try in the coming days to use those meta-classifiers relying on the sklearn api (so to speak) with RIPPER/IREP. 

Getting hands dirty is better than theorizing like I did in this issue. We'll see where it breaks, and perhaps I'll be even able to do a pull request, who knows ;-)

**Comment by imoscovitz** [comment]
Hi @flamby,

I'm sorry for the delay -- things have been really busy, but hoping they clear up in a couple of weeks or so. How's the metaclassification been working out for you?

That would be fantastic if you're interested in making a pull request! :)
